Min # Vending step9

undervineg opened the pull request.

undervine committed:
Add contents about feedback of step8

undervine committed:
Add and move extensions

undervine committed:
Make PieGraphView touchable and shakable

Override touches- functions and TouchStatus enum type to separate drawings by touch status.
Post notification to AdminViewController to request changing pieGraphView's size.
Override motionEnded() func resetting configurations to display initial pie graph.

undervine committed:
Add notification observer to detect graph size changes

Add resizePieGraph() fun to change pieGraphViews's bound size using it's graph(circle) size.
Make pieGraphView to become first responder when shake event occurs.

undervine committed:
Add contents about step9

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -9,88 +9,163 @@
         var startAngle: CGFloat = 0
         segments?.forEach { segment in
-            let endAngle: CGFloat = startAngle + radians(from: degree(of: segment.value))
+            let endAngle: CGFloat = startAngle + segment.value.degrees(in: sum).radians
```
draw(rect:) 함수는 뷰를 그리기 위해서 초당 60번까지도 반복해서 UIKit이 호출하는 메소드입니다.
따라서 이런 메소드에서 반복적으로 계산을 하거나 접근하는 메소드가 있다면 비효율적인 구조가 됩니다.
sum 같은 속성은 segments를 매번 reduce하고 있습니다. 초당 60번씩 호출할만한 값일까요? 
더구나 segments를 forEach로 반복하면서 접근하기 때문에 segments가 많을수록 느려질 것 같네요.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -9,88 +9,163 @@
+            NotificationCenter.default.post(
+                name: .didUpdatePieSize,
+                object: nil,
+                userInfo: [UserInfoKeys.newPieRadius: movedRadius ?? self.radius])
```
View가 노티를 보내서 결국 자기 자신의 크기를 바꾸는 구조가 적합할까요?
그냥 self.frame을 바꾸는 것과 어떤 차이가 있나요?

godrm commented on VendingMachineApp/VendingMachineApp/AdminViewController.swift:
```diff
@@ -27,13 +37,23 @@ class AdminViewController: UIViewController {
+        guard let userInfo = notification.userInfo else { return }
+        if let newPieRadius = userInfo[UserInfoKeys.newPieRadius] as? CGFloat {
+            pieGraphView.bounds.size.width = newPieRadius*2+6
+            pieGraphView.bounds.size.height = newPieRadius*2+6
```
UIView의 frame 과 bounds 차이점을 설명해보세요.
덧붙여서 frame을 바꾸지 않고 bounds를 바꾼 이유가 궁금합니다.

undervineg replied:
frame은 상위뷰의 좌표 시스템을 이용하고, bounds는 자기자신의 좌표 시스템을 사용합니다. 위 코드에서는 frame을 변경하면 화면에 이상하게 표시가 돼서 bounds를 썼는데요. 제 생각엔 frame은 상위 뷰에 상대적인 좌표를 가지고 있기 때문에 width나 height만 변경하면 좌상단 좌표는 고정인 채로 크기가 변경되는 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -9,88 +9,163 @@
+        case moved
+        case ended
+    }
+    var status: TouchStatus = .none
```
status 속성은 왜 private 접근 제어가 안되어 있는걸까요?

godrm commented on VendingMachineApp/VendingMachineApp/PieGraphView.swift:
```diff
@@ -9,88 +9,163 @@
+    required init?(coder aDecoder: NSCoder) {
+        self.initialFrameSize = CGSize()
+        super.init(coder: aDecoder)
+        self.initialFrameSize = frame.size
```
여기 코드는 왜 필요한지 이해하지 못하겠습니다. initialFrameSize 값은 어떤 용도로 언제 사용하는거죠?

undervineg replied:
쉐이크 시 기본 크기로 돌아가기 위해 파이 그래프 생성 시 초기 프레임 사이즈를 저장하려고 썼었습니다.

undervine committed:
[Tide Up] Extensions, Protocols 등 Utilities 폴더 내 파일 정리

undervine committed:
[Add] 구매이력 업데이트 시, 총 판매개수 계산

undervine committed:
[이동] TouchStatus 열거형을 PieGraphView 밖으로 분리

undervine committed:
[변경] AdminViewController에서 PieGraphView에 세그먼트 값들의 합(=총 판매개수)을 주입

[변경] PieGraphView에서 draw 시마다 계산했던 연산프로퍼티 sum을 저장프로퍼티로 변경

undervine committed:
[수정] 터치이벤트 시 뷰 크기를 PieGraphView 내부에서 변경하도록 수정

터치이벤트 moved 시 뷰 크기 변경.
뷰 크기가 변경되면 해당 뷰 크기의 bounds를 이용하여 원 그림.
터치이벤트 상태값을 UITouch.phase 프로퍼티를 사용하는 것으로 변경. 기존 TouchStatus 열거형 제거.

undervine committed:
[수정] 뷰컨트롤러에서 파이그래프 뷰에 데이터를 직접 주입하지 않고, DataSource 프로토콜을 사용

PieGraphDataSource 프로토콜 및 인터페이스 추가.
PieGraphView에 dataSource 프로퍼티 추가 및 인터페이스 활용.
AdminViewController에 인터페이스 구현 및 dataSource로 self 넘김.

undervineg commented:
- 전체 판매개수 데이터를 파이그래프에서 동적으로 계산하지 않고, 자판기->뷰컨트롤러->파이그래프로 전달하도록 변경
- 터치 이벤트 시, 뷰 크기를 파이그래프 내부에서만 변경하도록 수정
- TouchStatus 열거형 제거, UITouch 객체의 phase 프로퍼티 사용
- 뷰 컨트롤러에서 파이그래프 뷰의 데이터를 직접 변경하지 않고 프로토콜 인터페이스를 통해 넘김

godrm commented on VendingMachineApp/VendingMachineApp/Utilities/Protocols/PieGraphDataSource.swift:
```diff
@@ -0,0 +1,17 @@
+    func newSegment() -> Segment?
+
+    func totalValues() -> Int?
+}
```
데이터소스 프로토콜을 만든건 좋네요. 다음 사항이 궁금합니다.
- initialSegments()는 언제마다(계속해서) 불리나?
- newSegment()는 언제마다 불리나? (한번에 하나만 추가하고 그려고 추가해야 하나?)
- 뷰는 언제 다시 그릴지 판단하고 datasource 프로토콜 메소드를 호출하나?

undervineg replied:
- initialSegments() 호출시기: dataSource가 세팅된 직후 호출되는데요. dataSource는 AdminViewController가 화면에 표시될 때마다 세팅됩니다. 따라서 관리자 화면이 노출될 때마다 호출된다고 볼 수 있습니다.
- newSegment() 호출시기: 아 newSegment()가 필요가 없게 되었네요. 기존에는 segments 프로퍼티가 세팅될 때마다 호출했었는데요. 관리자 화면이 표시될 때마다 dataSource가 세팅되고, dataSource가 세팅될 때마다 segments가 세팅되는데, 이 때 이미 추가된 세그먼트까지 세팅이 되기 때문에 newSegments()는 필요가 없게 되었습니다.
- 뷰는 AdminViewController가 노출되면 dataSource를 다시 세팅하고 메소드를 이용하여 파이그래프 뷰를 갱신합니다.

godrm replied:
보통 DataSource를 지정하는 시점과 뷰를 다시 그리는 시점은 분리하는 게 좋습니다.
데이터소스가 있다고 해서 실제로 데이터가 없을수도 있고,
데이터소스에 데이터가 변경되면 다시 그리도록 알려줄 수 있어야 하기 때문입니다.
newSegment()를 변경할 껀가요?

undervineg replied:
음... 데이터소스 지정 시점과 뷰를 다시 그리는 시점 분리를 더이상 어떻게 해야할 지 감이 안 잡혀요 ㅠ
pieGraphView는 AdminViewController가 viewDidLoad 될때랑 touch 이벤트 발생 시 다시 그리고, dataSource는 AdminViewController의 viewDidLoad 시점에 주입하도록 했습니다. pieGraphView는 dataSource가 세팅된 직후 dataSource의 인터페이스를 사용하여 필요한 데이터를 세팅하도록 했습니다. 불필요한 setNeedsDisplay는 지웠습니다. 어느 부분을 더 보면 될까요?

godrm replied:
이럴때 각자 객체 입장에서 생각해볼 필요가 있습니다.
- DataSource는 언제 데이터가 초기화되고, 데이터가 갱신되나, 그래서 언제 다시 그리라고 요청해야 하나, 전체를 다 갱신해야 하나 일부만 갱신해야 하나...
- View는 언제 DataSource가 설정되고, 언제 데이터를 확인해서 그릴까, 전체를 갱신하는 시점과 일부만 갱신하는 시점 동작 구분이 필요한가...
아직 해답을 모른다고 해도 지금도 나쁘지 않습니다. 테이블뷰에 대해 공부해보면 조금 힌트가 생길겁니다.

undervineg replied:
네! 감사합니다.

undervine committed:
Remove newSegment() from PieGraphDataSource protocol

undervine committed:
[수정] dataSource 세팅되면 segments, sum 데이터를 세팅하도록 수정

undervine committed:
[수정] pieGraphView의 dataSource 세팅 위치를 viewDidLoad()로 변경

godrm commented:
수고하셨습니다. 이제 다음 미션으로 갑시다. 
DataSource에 동작에 대해서는 다다음 테이블뷰를 보면 좀 더 이해가 될겁니다.
